Docs: Grammar action, satisfaction! First 20 pages. (#1817)
dlt-hub/dlt: docs/website/docs/general-usage/pipeline.md

# Pipeline

A [pipeline](glossary.md#pipeline) is a connection that moves data from your Python code to a
[destination](glossary.md#destination). The pipeline accepts `dlt` [sources](source.md) or
[resources](resource.md), as well as generators, async generators, lists, and any iterables.
Once the pipeline runs, all resources are evaluated and the data is loaded at the destination.

Example:

> This pipeline will load a list of objects into a `duckdb` table named "three":

Reviewer: ```suggestion
This pipeline will load a list of objects into a DuckDB table named "three":
```

print(info)
```

You instantiate a pipeline by calling the `dlt.pipeline` function with the following arguments:

- `pipeline_name`: a name of the pipeline that will be used to identify it in trace and monitoring
  events and to restore its state and data schemas on subsequent runs. If not provided, `dlt` will
  create a pipeline name from the file name of the currently executing Python module.
- `destination`: a name of the [destination](../dlt-ecosystem/destinations) to which dlt
  will load the data. It may also be provided to the `run` method of the `pipeline`.
- `dataset_name`: a name of the dataset to which the data will be loaded. A dataset is a logical
  group of tables, i.e., `schema` in relational databases or a folder grouping many files. It may also be
  provided later to the `run` or `load` methods of the pipeline. If not provided at all, then
  it defaults to the `pipeline_name`.

To load the data, you call the `run` method and pass your data in the `data` argument.

Arguments:

> - `data` (the first argument) may be a dlt source, resource, generator function, or any Iterator /

Reviewer: ```suggestion
- `data` (the first argument) may be a dlt source, resource, generator function, or any Iterator or
```

import dlt
from dlt.common.pipeline import get_dlt_pipelines_dir

dev_pipelines_dir = os.path.join(get_dlt_pipelines_dir(), "dev")
pipeline = dlt.pipeline(destination="duckdb", dataset_name="sequence", pipelines_dir=dev_pipelines_dir)
```
This code stores the pipeline working folder in `~/.dlt/pipelines/dev/<pipeline_name>`. Note that you need to pass this `~/.dlt/pipelines/dev/`
into all CLI commands to get info/trace for that pipeline.

## Do experiments with dev mode

If you [create a new pipeline script](../walkthroughs/create-a-pipeline.md), you will be
experimenting a lot. If you want each time the pipeline resets its state and loads data to a
new dataset, set the `dev_mode` argument of the `dlt.pipeline` method to True. Each time the
pipeline is created, `dlt` adds a datetime-based suffix to the dataset name.

## Refresh pipeline data and state

You can reset parts or all of your sources by using the `refresh` argument to `dlt.pipeline` or the pipeline's `run` or `extract` method.
That means when you run the pipeline, the sources/resources being processed will have their state reset and their tables either dropped or truncated,
depending on which refresh mode is used.

> The `refresh` option works with all relational/SQL destinations and file buckets (`filesystem`). It does not work with vector databases (we are working on that) and

Reviewer: ```suggestion
The `refresh` option works with all relational or SQL destinations and cloud storages and files (`filesystem`). It does not work with vector databases (we are working on that) and
```

pipeline = dlt.pipeline("airtable_demo", destination="duckdb")
pipeline.run(airtable_emojis().with_resources("📆 Schedule"), refresh="drop_data")
```
Above, the incremental state of the "📆 Schedule" is reset before the `extract` step so data is fully reacquired. Just before the `load` step starts,
the "_schedule" is truncated, and new (full) table data will be inserted/copied.

## Display the loading progress

You can add a progress monitor to the pipeline. Typically, its role is to visually assure the user that

> the pipeline run is progressing. `dlt` supports 4 progress monitors out of the box:

Reviewer: ```suggestion
the pipeline run is progressing. dlt supports 4 progress monitors out of the box:
```
